src/Security/TOTPManager.php

```diff
@@ -0,0 +1,379 @@
+<?php
+
+/**
+ * ---------------------------------------------------------------------
+ *
+ * GLPI - Gestionnaire Libre de Parc Informatique
+ *
+ * http://glpi-project.org
+ *
+ * @copyright 2015-2022 Teclib' and contributors.
+ * @copyright 2003-2014 by the INDEPNET Development Team.
+ * @licence   https://www.gnu.org/licenses/gpl-3.0.html
+ *
+ * ---------------------------------------------------------------------
+ *
+ * LICENSE
+ *
+ * This file is part of GLPI.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * ---------------------------------------------------------------------
+ */
+namespace Glpi\Security;
+
+use Glpi\Application\View\TemplateRenderer;
+use RobThree\Auth\Providers\Qr\BaconQrCodeProvider;
+use RobThree\Auth\TwoFactorAuth;
+
+final class TOTPManager
+{
+    public const SECRET_LENGTH_BITS = 160;
+
+    public const CODE_LENGTH_DIGITS = 6;
+
+    public const CODE_VALIDITY_SECONDS = 30;
+
+    public const CODE_ALGORITHM = 'sha1';
```

Comment:
SHA1 should not be avoided? since 2015, it is an algorithm that is considered insecure.

Reply:
Yes it should be. However, Google Authenticator doesn't work when I tried SHA-256. I spent a lot of time trying to figure out why it wasn't working with it.

Reply:
An important note is that the RFC for TOTP (rfc6238) is based on HOTP (rfc4226) which specifically mentions SHA-1. By default, authenticator apps seem to use SHA-1 by default for compatibility reasons. When you use a QR code instead of a manually entered secret, it is actually a `otpauth` URI which can contain an `algorithm` parameter to specify the use of a different algorithm. For some reason, the Google Authenticator still ignores the parameter in the URI (I verified it was set correctly). I don't know why we should expect them to support a parameter of a URI they formalized themselves...
https://github.com/google/google-authenticator/wiki/Key-Uri-Format#algorithm

As for the real world implications, there are people who know way more than me on the subject. However, from what I understand, TOTP uses HMAC (HMAC-SHA-1 in this case) which does reduce the risks of using SHA-1 as the secret key is assumed to be unknown by anyone except the application and client.